File: server/internal/domain/top_up.go
package domain

import "time"

type TopUpStatus int

const (
	TopUpPending  TopUpStatus = 0
	TopUpSuccess  TopUpStatus = 1
	TopUpFailed   TopUpStatus = 2
	TopUpRefunded TopUpStatus = 3
)

type TopUp struct {
	ID            uint64      `gorm:"primaryKey" json:"id"`
	UserID        uint64      `gorm:"not null;index" json:"user_id"`
	PackageID     *uint64     `gorm:"index" json:"package_id,omitempty"`
	AmountCents   int64       `gorm:"not null" json:"amount_cents"`
	QuotaGranted  int64       `gorm:"not null" json:"quota_granted"`
	Currency      string      `gorm:"size:8;not null;default:'CNY'" json:"currency"`
	PaymentMethod string      `gorm:"size:32;not null" json:"payment_method"`
	PaymentStatus TopUpStatus `gorm:"not null;default:0" json:"payment_status"`
	TransactionID *string     `gorm:"size:255" json:"transaction_id,omitempty"`
	PaidAt        *time.Time  `json:"paid_at,omitempty"`
	CreatedAt     time.Time   `json:"created_at"`
	UpdatedAt     time.Time   `json:"updated_at"`
}

func (TopUp) TableName() string { return "top_ups" }

File: admin/src/api/topup.ts
import { client } from './client'
import type { ApiResponse, PagedResponse } from '../types/api'

export interface TopUp {
  id: number
  user_id: number
  package_id?: number
  amount_cents: number
  quota_granted: number
  currency: string
  payment_method: string
  payment_status: number
  transaction_id?: string
  paid_at?: string
  created_at: string
  updated_at: string
}

export interface TopUpForm {
  user_id: number
  quota_granted: number
}

export function listTopUps(
  page = 1,
  pageSize = 20,
  userId?: number,
  status?: number,
  startDate?: string,
  endDate?: string,
) {
  let url = `/topups?page=${page}&page_size=${pageSize}`
  if (userId) {
    url += `&user_id=${userId}`
  }
  if (status !== undefined) {
    url += `&status=${status}`
  }
  if (startDate) {
    url += `&start_date=${startDate}`
  }
  if (endDate) {
    url += `&end_date=${endDate}`
  }
  return client.get<ApiResponse<PagedResponse<TopUp>>, ApiResponse<PagedResponse<TopUp>>>(url)
}

export function createTopUp(data: TopUpForm) {
  return client.post<ApiResponse<TopUp>, ApiResponse<TopUp>>('/topups', data)
}

export function markTopUpPaid(id: number, transactionId = '') {
  return client.post<ApiResponse<TopUp>, ApiResponse<TopUp>>(
    `/topups/${id}/paid?transaction_id=${encodeURIComponent(transactionId)}`,
  )
}

export function markTopUpFailed(id: number) {
  return client.post<ApiResponse<TopUp>, ApiResponse<TopUp>>(`/topups/${id}/failed`)
}

export function refundTopUp(id: number) {
  return client.post<ApiResponse<TopUp>, ApiResponse<TopUp>>(`/topups/${id}/refund`)
}

// Reserved for future use
// export function getTopUp(id: number) {
//   return client.get<ApiResponse<TopUp>, ApiResponse<TopUp>>(`/topups/${id}`)
// }

export function batchUpdateTopUpStatus(ids: number[], status: string) {
  return client.post<{ ids: number[]; status: string }, ApiResponse<{ affected: number }>>(
    '/topups/batch-status',
    { ids, status },
  )
}
